File: Problems/Heaps/Merge-k-Sorted-Lists/Merge-k-Sorted-Lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
**/

public class Solution
{
    public ListNode MergeKLists(ListNode[] lists)
    {
        var dummy = new ListNode(0);
        var tail = dummy;

        var pq = new PriorityQueue<ListNode, int>(lists.Where(x => x != null).Select(x => (x, x.val)));

        while (pq.Count > 0)
        {
            var node = pq.Dequeue();

            if (node.next != null)
            {
                pq.Enqueue(node.next, node.next.val);
            }

            tail.next = node;
            tail = node;
        }

        return dummy.next;
    }
}

File: Problems/Heaps/Merge-k-Sorted-Lists/Merge-k-Sorted-Lists.cpp
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode dummy, *tail = &dummy;
        auto cmp = [](auto a, auto b) { return a->val > b->val; };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> q(cmp);
        for (auto list : lists) {
            if (list) q.push(list); // avoid pushing NULL list.
        }
        while (q.size()) {
            auto node = q.top();
            q.pop();
            if (node->next) q.push(node->next);
            tail->next = node;
            tail = node;
        }
        return dummy.next;
    }
};
